File: hearts-scala/src/main/scala/funar/hearts/Player.scala
package funar.hearts

import cats.data._
import org.atnos.eff._
import org.atnos.eff.all._
import org.atnos.eff.syntax.all._

case class Player(id: String, name: String) {
  override def equals(other: Any): Boolean =
    other match {
      case other: Player => this.id == other.id
      case _ => false
    }

  override def hashCode: Int = this.id.hashCode
}


object Player {
  import GameEvent._

  type Pile = Set[Card]
  case class PlayerState[PlayerId](hand: Hand, trick: Trick, pile: Pile)

  def emptyPlayerState[PlayerId] = PlayerState[PlayerId](Hand.empty, Trick.empty, Set.empty)

  def playerProcessEvent[PlayerId](player: Player, event: GameEvent, state: PlayerState[PlayerId]): PlayerState[PlayerId] =
    event match {
      case HandDealt(player1, hand) =>
        if (player == player1)
          PlayerState(hand = hand, trick = Trick.empty, pile = Set.empty)
        else
          state
      case PlayerTurnChanged(player1) => state
      case LegalCardPlayed(player1, card) =>
        if (player == player1)
          state.copy(hand = state.hand - card,
                     trick = Trick.add(state.trick, player1, card))
        else
          state.copy(trick = Trick.add(state.trick, player1, card))
      case IllegalCardPlayed(player1, card) => state
      case TrickTaken(player1, trick) =>
        if (player1 == player)
          state.copy(trick = Trick.empty,
                     pile = state.pile ++ Trick.cards(trick))
        else
          state.copy(trick = Trick.empty)
      case GameEnded(winner) => state
    }

  def playerProcessEventM[PlayerId, Effects](player: Player, event: GameEvent)
          (implicit member: State[PlayerState[PlayerId], _] |= Effects): Eff[Effects, Unit] =
    for {
      playerState <- get[Effects, PlayerState[PlayerId]]
      playerState1 = playerProcessEvent(player, event, playerState)
      _ <- put[Effects, PlayerState[PlayerId]](playerState1)
    } yield ()

  type EventProcessor[Effects, Event, Command] = Event => Eff[Effects, Seq[Command]]

  type Strategy[Effects] = Player => EventProcessor[Effects, GameEvent, GameCommand]

  type Chooser[PlayerId, Effects] = (Player, PlayerState[PlayerId]) => Eff[Effects, Card]

  def chooserStrategy[PlayerId, Effects](chooser: Chooser[PlayerId, Effects])
          (implicit member: State[PlayerState[PlayerId], _] |= Effects): Strategy[Effects] = {
    import GameCommand._
    (player: Player) => 
      (event: GameEvent) =>
        for {
          _ <- playerProcessEventM(player, event)
          playerState <- get[Effects, PlayerState[PlayerId]]
          commands <- event match {
            case HandDealt(player1, hand) => {
              println("player " ++ player.toString ++ " received HandDealt for player " ++ player1.toString ++ " (" ++ (player == player1).toString ++ ")")
              if ((player == player1) && hand.contains(Card(Suit.Clubs, Rank.Two)))
                Eff.pure[Effects, Seq[GameCommand]](Seq(PlayCard(player, Card(Suit.Clubs, Rank.Two))))
              else
                Eff.pure[Effects, Seq[GameCommand]](Seq.empty)
            }
            case PlayerTurnChanged(player1) =>
              if (player == player1)
                for { card <- chooser(player, playerState) } yield Seq(PlayCard(player, card))
              else
                Eff.pure[Effects, Seq[GameCommand]](Seq.empty)
            case LegalCardPlayed(player1, card) => Eff.pure[Effects, Seq[GameCommand]](Seq.empty)
            case IllegalCardPlayed(player1, card) => Eff.pure[Effects, Seq[GameCommand]](Seq.empty)
            case TrickTaken(player1, trick) => Eff.pure[Effects, Seq[GameCommand]](Seq.empty)
            case GameEnded(winner) => Eff.pure[Effects, Seq[GameCommand]](Seq.empty)
          }

        } yield commands
  }

  def chooseAlong[PlayerId, Effects]: Chooser[PlayerId, Effects] = {
    (_, playerState) =>
    if (Trick.isEmpty(playerState.trick))
      Eff.pure[Effects, Card](playerState.hand.minBy(_.rank))  
    else {
      val firstCard = Trick.leadingCard(playerState.trick)
      val followingCardsOnHand = playerState.hand.filter { _.suit == firstCard.suit }
      followingCardsOnHand.minByOption(_.rank) match {
        case None => Eff.pure[Effects, Card](playerState.hand.maxBy(_.rank))
        case Some(card) => Eff.pure[Effects, Card](card)
      }
    } 
  }

  def alongStrategy[PlayerId, Effects](implicit member: State[PlayerState[PlayerId], _] |= Effects): Strategy[Effects] =
    chooserStrategy(chooseAlong[PlayerId, Effects])

  def chooseInteractive[Effects, PlayerId](implicit member: Teletype |= Effects): Chooser[PlayerId, Effects] = {
    (player, playerState) =>
      for {
        _ <- Teletype.writeTTY("Your turn, player " ++ player.name)
        handList = playerState.hand.toList
        _ <- if (Trick.isEmpty(playerState.trick))
                Teletype.writeTTY("You lead the next trick.")
             else
                Teletype.writeTTY("Trick: " ++ Trick.pretty(playerState.trick))
        _ <- Teletype.writeTTY("Your hand:\n" ++ prettyCards(handList))
        ncards = playerState.hand.size
        _ <- Teletype.writeTTY("Pick a card (1-"  ++ ncards.toString ++ ")")
        selected <- getNumber(1, ncards)
      } yield handList(selected-1)
  }

  def getNumber[Effects](lo: Int, hi: Int)(implicit member: Teletype |= Effects): Eff[Effects, Int] =
    for {
      s <- Teletype.readTTY
      input = s.toInt
      n <- if ((lo <= input) && (input <= hi))
             Eff.pure[Effects, Int](input)
           else
            for {
              _ <- Teletype.writeTTY("Input must be between " ++ lo.toString ++ " and " ++ hi.toString ++ ". Try again.")
              n <- getNumber(lo, hi)
            } yield n
    } yield n

  def prettyCards(cards: List[Card]): String = {
    def prettyOne(n: Integer, card: Card): String = "(" ++ n.toString ++ ")" ++ Trick.prettyCard(card)
    def loop(n: Integer, cards: List[Card]): String =
      cards match {
        case Nil => ""
        case List(card) => prettyOne(n, card)
        case card::cards => prettyOne(n, card) ++ "\n" ++ loop(n+1, cards)
      }
    loop(1, cards)
  }

  def interactiveStrategy[PlayerId, Effects]
    (implicit ttyState: Teletype |= Effects,
              memberState: State[PlayerState[PlayerId], _] |= Effects): Strategy[Effects] =
    chooserStrategy(chooseInteractive)


  
}

File: hearts-mike/src/Player.hs
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE GADTs, FlexibleContexts, TypeOperators, DataKinds, PolyKinds #-}
{-# LANGUAGE LambdaCase, BlockArguments, TypeApplications, ScopedTypeVariables, ConstraintKinds #-}
{-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}

module Player where

import qualified Data.Set as Set
import Data.Set (Set)
import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map, (!))

import Cards
import GameEvent

import Polysemy
import qualified Polysemy.State as State
import Polysemy.State (State)

import qualified Teletype
import Teletype (Teletype)

-- import Polysemy.Bundle

-- Spieler kennt das, was er sieht
data PlayerState player =
  PlayerState { playerStateHand  :: Hand,
                playerStateTrick :: Trick,
                playerStatePile :: [Card] }
  deriving Show

emptyPlayerState = PlayerState emptyHand [] []

-- Spielevent "sehen"
playerProcessEvent :: Player -> GameEvent -> PlayerState player -> PlayerState player
playerProcessEvent player (HandDealt player' hand) state
  | player == player' =
    PlayerState { playerStateHand = hand,
                  playerStateTrick = emptyTrick,
                  playerStatePile = [] }
  | otherwise = state
playerProcessEvent player (PlayerTurnChanged player') state = state
playerProcessEvent player (LegalCardPlayed player' card) state
  | player' == player =
    state { playerStateHand = removeCard card (playerStateHand state),
            playerStateTrick = addToTrick player' card (playerStateTrick state) }
  | otherwise =
    state { playerStateTrick = addToTrick player' card (playerStateTrick state) }
playerProcessEvent player (TrickTaken player' trick) state
  | player' == player =
    state { playerStateTrick = emptyTrick,
            playerStatePile = (cardsOfTrick trick) ++ (playerStatePile state) }
  | otherwise =
    state { playerStateTrick = emptyTrick }
playerProcessEvent player (IllegalCardAttempted player' card) state = state
playerProcessEvent player (GameEnded winner) state = state

-- monadische Version
playerProcessEventM ::
  Member (State (PlayerState player)) effects => Player -> GameEvent -> Sem effects ()
playerProcessEventM player event =
  do playerState <- State.get
     let playerState' = playerProcessEvent player event playerState
     State.put playerState'

type GameEventProcessor effects = GameEvent -> Sem effects [GameCommand]

-- ... macht aus Events Commands
type Strategy effects = Player -> GameEventProcessor effects

type Chooser player effects =
  Player -> PlayerState player -> Sem effects Card

-- Das hier ist schade, weil es bedeutet, daß PlayerState player bereits
-- in den Effekten des Chooser stehen muß.
-- Aber sonst wird es später schwierig, die entstehenden Strategien
-- in einer Monade zu kombinieren.
chooserStrategy ::
  Member (State (PlayerState player)) effects => Chooser player effects -> Strategy effects
chooserStrategy choose =
  \ player event ->
    do playerProcessEventM player event
       playerState <- State.get
       case event of
         HandDealt player' hand ->
           if (player == player') && (containsCard (Card Clubs Two) hand)
           then return [PlayCard player (Card Clubs Two)]
           else return []
         PlayerTurnChanged player' ->
           if player == player'
           then do card <- choose player playerState
                   return [PlayCard player card]
           else return []
         LegalCardPlayed player' card -> return []
         IllegalCardAttempted player' card -> return []
         TrickTaken player' trick -> return []
         GameEnded winner -> return []

-- Strategie für Roboterspieler
chooseAlong :: Chooser player effects
chooseAlong _ playerState =
  case playerStateTrick playerState of
    [] -> return (Set.findMin (playerStateHand playerState))       -- leine erste Karte
    trick ->
      let hand = playerStateHand playerState
          (_, firstCard) = last trick
          firstSuit = suit firstCard
          followingCardsOnHand = Set.filter ((== firstSuit) . suit) hand
      in case Set.lookupMin followingCardsOnHand of
           Nothing ->
             return (Set.findMax hand) -- wir haben nix passendes, nimm große Karte
           Just card ->
             return card           -- sonst kleine passende

alongStrategy :: forall player effects . Member (State (PlayerState player)) effects => Strategy effects
alongStrategy player event = -- braucht Eta-Expansion
  chooserStrategy chooseAlong player event

-- newtype InteractiveEffect player monad a = InteractiveEffect (Bundle '[State (PlayerState player), Teletype] monad a)

-- interaktiver Spieler
chooseInteractive :: forall player effects . Member Teletype effects => Chooser player effects
chooseInteractive player playerState =
  do Teletype.writeTTY ("Your turn, player " ++ (playerName player))
     case playerStateTrick playerState of
       [] ->
         Teletype.writeTTY "You lead the next trick."
       trick ->
         Teletype.writeTTY ("Trick: " ++ prettyTrick trick)
     let hand = playerStateHand playerState
         handList = Set.elems hand
         ncards = Set.size hand
     Teletype.writeTTY ("Your hand:\n" ++ prettyCards handList)
     Teletype.writeTTY ("Pick a card (1-" ++ show ncards ++ ")")
     selected <- getNumber (1,ncards)
     return (handList !! (selected - 1))

prettyTrick trick =
  let prettyOne (player, card) = prettyCard card ++ " from " ++ prettyPlayer player
      pretty [] = ""
      pretty [one] = prettyOne one
      pretty (one:rest) = prettyOne one ++ ", " ++ pretty rest
  in pretty (reverse trick)

-- Spieleridentität drucken
prettyPlayer (Player id name) = name ++ "(" ++ id ++ ")"


-- |Eine Karte ausdrucken
prettyCard :: Card -> String
prettyCard c = show (rank c) ++ " of " ++ show (suit c)

-- |Liste von Karten ausdrucken
prettyCards :: [Card] -> String
prettyCards cards =
  let loop _ [] = ""
      loop n [card] = prettyOne n card
      loop n (card:cards) = prettyOne n card ++ "\n" ++ loop (n+1) cards
      prettyOne n card = "(" ++ (show n) ++ ") " ++ prettyCard card
  in loop 1 cards

-- Zahl einlesen
getNumber :: (Num a, Ord a, Read a, Show a, Member Teletype effects) => (a, a) -> Sem effects a
getNumber (lo, hi) = do
  s <- Teletype.readTTY
  let input = read s
  if lo <= input && input <= hi
  then return input
  else
    do Teletype.writeTTY ("Input must be between " ++ (show lo) ++ " and " ++ (show hi) ++ ". Try again")
       getNumber (lo, hi)

interactiveStrategy :: forall player effects . (Member Teletype effects, Member (State (PlayerState player)) effects) => Strategy effects
interactiveStrategy player event =
  chooserStrategy chooseInteractive player event

